File: julia/Tensor4all.jl/src/MPO/MPO.jl
"""
    MPO

Submodule for MPO (Matrix Product Operator) operations (corresponds to tensor4all-mpocontraction).

# Basic Usage

```julia
using Tensor4all.MPO

# Create an identity MPO
mpo = identity(Float64, [2, 3])

# Contract two MPOs
result = contract(mpo_a, mpo_b; algorithm=:zipup, tolerance=1e-12)
```
"""
module MPO

using Libdl

# Import parent module's library management
import ..get_lib, ..C_API, ..Algorithm

# Use C_API's status code
const T4A_SUCCESS = C_API.T4A_SUCCESS

# ============================================================================
# Contraction Algorithm Enum (from Algorithm module)
# ============================================================================

# Contraction algorithm values (must match Rust t4a_contraction_algorithm)
const CONTRACTION_NAIVE = Cint(0)
const CONTRACTION_ZIPUP = Cint(1)
const CONTRACTION_FIT = Cint(2)

"""
    algorithm_to_cint(alg::Symbol) -> Cint

Convert Julia symbol to C API contraction algorithm integer.
"""
function algorithm_to_cint(alg::Symbol)
    if alg == :naive
        return CONTRACTION_NAIVE
    elseif alg == :zipup
        return CONTRACTION_ZIPUP
    elseif alg == :fit
        return CONTRACTION_FIT
    else
        error("Unknown contraction algorithm: $alg. Use :naive, :zipup, or :fit")
    end
end

# ============================================================================
# MPOF64
# ============================================================================

"""
    MPOF64

A Matrix Product Operator with Float64 elements.

Wraps the Rust `MPO<f64>` type via the C API.
"""
mutable struct MPOF64
    ptr::Ptr{Cvoid}

    function MPOF64(ptr::Ptr{Cvoid})
        ptr == C_NULL && error("Failed to create MPOF64 (null pointer)")
        mpo = new(ptr)
        finalizer(_release_mpo_f64, mpo)
        return mpo
    end
end

function _release_mpo_f64(mpo::MPOF64)
    if mpo.ptr != C_NULL
        lib = get_lib()
        ccall(Libdl.dlsym(lib, :t4a_mpo_f64_release), Cvoid, (Ptr{Cvoid},), mpo.ptr)
        mpo.ptr = C_NULL
    end
end

# ============================================================================
# MPOC64 (Complex{Float64})
# ============================================================================

"""
    MPOC64

A Matrix Product Operator with Complex{Float64} elements.

Wraps the Rust `MPO<Complex64>` type via the C API.
"""
mutable struct MPOC64
    ptr::Ptr{Cvoid}

    function MPOC64(ptr::Ptr{Cvoid})
        ptr == C_NULL && error("Failed to create MPOC64 (null pointer)")
        mpo = new(ptr)
        finalizer(_release_mpo_c64, mpo)
        return mpo
    end
end

function _release_mpo_c64(mpo::MPOC64)
    if mpo.ptr != C_NULL
        lib = get_lib()
        ccall(Libdl.dlsym(lib, :t4a_mpo_c64_release), Cvoid, (Ptr{Cvoid},), mpo.ptr)
        mpo.ptr = C_NULL
    end
end

# ============================================================================
# Constructors - Float64
# ============================================================================

"""
    zeros(::Type{Float64}, site_dims::Vector{Tuple{Int,Int}})

Create an MPO representing the zero operator.
"""
function zeros(::Type{Float64}, site_dims::Vector{Tuple{Int,Int}})
    lib = get_lib()
    dims1 = convert(Vector{Csize_t}, [d[1] for d in site_dims])
    dims2 = convert(Vector{Csize_t}, [d[2] for d in site_dims])
    ptr = ccall(
        Libdl.dlsym(lib, :t4a_mpo_f64_new_zeros),
        Ptr{Cvoid},
        (Ptr{Csize_t}, Ptr{Csize_t}, Csize_t),
        dims1,
        dims2,
        length(site_dims),
    )
    return MPOF64(ptr)
end

"""
    constant(::Type{Float64}, site_dims::Vector{Tuple{Int,Int}}, value::Float64)

Create an MPO representing a constant operator.
"""
function constant(::Type{Float64}, site_dims::Vector{Tuple{Int,Int}}, value::Float64)
    lib = get_lib()
    dims1 = convert(Vector{Csize_t}, [d[1] for d in site_dims])
    dims2 = convert(Vector{Csize_t}, [d[2] for d in site_dims])
    ptr = ccall(
        Libdl.dlsym(lib, :t4a_mpo_f64_new_constant),
        Ptr{Cvoid},
        (Ptr{Csize_t}, Ptr{Csize_t}, Csize_t, Cdouble),
        dims1,
        dims2,
        length(site_dims),
        value,
    )
    return MPOF64(ptr)
end

"""
    identity(::Type{Float64}, site_dims::Vector{Int})

Create an identity MPO.
"""
function identity(::Type{Float64}, site_dims::Vector{Int})
    lib = get_lib()
    dims = convert(Vector{Csize_t}, site_dims)
    ptr = ccall(
        Libdl.dlsym(lib, :t4a_mpo_f64_new_identity),
        Ptr{Cvoid},
        (Ptr{Csize_t}, Csize_t),
        dims,
        length(dims),
    )
    return MPOF64(ptr)
end

# ============================================================================
# Constructors - ComplexF64
# ============================================================================

"""
    zeros(::Type{ComplexF64}, site_dims::Vector{Tuple{Int,Int}})

Create a complex MPO representing the zero operator.
"""
function zeros(::Type{ComplexF64}, site_dims::Vector{Tuple{Int,Int}})
    lib = get_lib()
    dims1 = convert(Vector{Csize_t}, [d[1] for d in site_dims])
    dims2 = convert(Vector{Csize_t}, [d[2] for d in site_dims])
    ptr = ccall(
        Libdl.dlsym(lib, :t4a_mpo_c64_new_zeros),
        Ptr{Cvoid},
        (Ptr{Csize_t}, Ptr{Csize_t}, Csize_t),
        dims1,
        dims2,
        length(site_dims),
    )
    return MPOC64(ptr)
end

"""
    constant(::Type{ComplexF64}, site_dims::Vector{Tuple{Int,Int}}, value::ComplexF64)

Create a complex MPO representing a constant operator.
"""
function constant(::Type{ComplexF64}, site_dims::Vector{Tuple{Int,Int}}, value::ComplexF64)
    lib = get_lib()
    dims1 = convert(Vector{Csize_t}, [d[1] for d in site_dims])
    dims2 = convert(Vector{Csize_t}, [d[2] for d in site_dims])
    ptr = ccall(
        Libdl.dlsym(lib, :t4a_mpo_c64_new_constant),
        Ptr{Cvoid},
        (Ptr{Csize_t}, Ptr{Csize_t}, Csize_t, Cdouble, Cdouble),
        dims1,
        dims2,
        length(site_dims),
        real(value),
        imag(value),
    )
    return MPOC64(ptr)
end

"""
    identity(::Type{ComplexF64}, site_dims::Vector{Int})

Create a complex identity MPO.
"""
function identity(::Type{ComplexF64}, site_dims::Vector{Int})
    lib = get_lib()
    dims = convert(Vector{Csize_t}, site_dims)
    ptr = ccall(
        Libdl.dlsym(lib, :t4a_mpo_c64_new_identity),
        Ptr{Cvoid},
        (Ptr{Csize_t}, Csize_t),
        dims,
        length(dims),
    )
    return MPOC64(ptr)
end

# ============================================================================
# Properties - Float64
# ============================================================================

function Base.length(mpo::MPOF64)
    lib = get_lib()
    out_len = Ref{Csize_t}(0)
    status = ccall(
        Libdl.dlsym(lib, :t4a_mpo_f64_len),
        Cint,
        (Ptr{Cvoid}, Ptr{Csize_t}),
        mpo.ptr,
        out_len,
    )
    status == T4A_SUCCESS || error("Failed to get length: status = $status")
    return Int(out_len[])
end

"""
    rank(mpo::MPOF64)

Get the maximum bond dimension (rank) of the MPO.
"""
function rank(mpo::MPOF64)
    lib = get_lib()
    out_rank = Ref{Csize_t}(0)
    status = ccall(
        Libdl.dlsym(lib, :t4a_mpo_f64_rank),
        Cint,
        (Ptr{Cvoid}, Ptr{Csize_t}),
        mpo.ptr,
        out_rank,
    )
    status == T4A_SUCCESS || error("Failed to get rank: status = $status")
    return Int(out_rank[])
end

# ============================================================================
# Properties - ComplexF64
# ============================================================================

function Base.length(mpo::MPOC64)
    lib = get_lib()
    out_len = Ref{Csize_t}(0)
    status = ccall(
        Libdl.dlsym(lib, :t4a_mpo_c64_len),
        Cint,
        (Ptr{Cvoid}, Ptr{Csize_t}),
        mpo.ptr,
        out_len,
    )
    status == T4A_SUCCESS || error("Failed to get length: status = $status")
    return Int(out_len[])
end

# ============================================================================
# Contraction - Float64
# ============================================================================

"""
    contract_naive(a::MPOF64, b::MPOF64)

Contract two MPOs using naive (exact) algorithm.
"""
function contract_naive(a::MPOF64, b::MPOF64)
    lib = get_lib()
    ptr = ccall(
        Libdl.dlsym(lib, :t4a_mpo_f64_contract_naive),
        Ptr{Cvoid},
        (Ptr{Cvoid}, Ptr{Cvoid}),
        a.ptr,
        b.ptr,
    )
    return MPOF64(ptr)
end

"""
    contract_zipup(a::MPOF64, b::MPOF64; tolerance=1e-12, max_bond_dim=0)

Contract two MPOs using zip-up algorithm with compression.
"""
function contract_zipup(
    a::MPOF64,
    b::MPOF64;
    tolerance::Float64 = 1e-12,
    max_bond_dim::Int = 0,
)
    lib = get_lib()
    ptr = ccall(
        Libdl.dlsym(lib, :t4a_mpo_f64_contract_zipup),
        Ptr{Cvoid},
        (Ptr{Cvoid}, Ptr{Cvoid}, Cdouble, Csize_t),
        a.ptr,
        b.ptr,
        tolerance,
        max_bond_dim,
    )
    return MPOF64(ptr)
end

"""
    contract_fit(a::MPOF64, b::MPOF64; tolerance=1e-12, max_bond_dim=0, max_sweeps=0)

Contract two MPOs using variational fitting algorithm.
"""
function contract_fit(
    a::MPOF64,
    b::MPOF64;
    tolerance::Float64 = 1e-12,
    max_bond_dim::Int = 0,
    max_sweeps::Int = 0,
)
    lib = get_lib()
    ptr = ccall(
        Libdl.dlsym(lib, :t4a_mpo_f64_contract_fit),
        Ptr{Cvoid},
        (Ptr{Cvoid}, Ptr{Cvoid}, Cdouble, Csize_t, Csize_t),
        a.ptr,
        b.ptr,
        tolerance,
        max_bond_dim,
        max_sweeps,
    )
    return MPOF64(ptr)
end

"""
    contract(a::MPOF64, b::MPOF64; algorithm=:naive, tolerance=1e-12, max_bond_dim=0)

Contract two MPOs using the specified algorithm.

# Arguments
- `a, b`: MPOs to contract
- `algorithm`: Contraction algorithm (:naive, :zipup, or :fit)
- `tolerance`: Relative tolerance for truncation
- `max_bond_dim`: Maximum bond dimension (0 for unlimited)
"""
function contract(
    a::MPOF64,
    b::MPOF64;
    algorithm::Symbol = :naive,
    tolerance::Float64 = 1e-12,
    max_bond_dim::Int = 0,
)
    lib = get_lib()
    alg_cint = algorithm_to_cint(algorithm)
    ptr = ccall(
        Libdl.dlsym(lib, :t4a_mpo_f64_contract),
        Ptr{Cvoid},
        (Ptr{Cvoid}, Ptr{Cvoid}, Cint, Cdouble, Csize_t),
        a.ptr,
        b.ptr,
        alg_cint,
        tolerance,
        max_bond_dim,
    )
    return MPOF64(ptr)
end

# ============================================================================
# Contraction - ComplexF64
# ============================================================================

"""
    contract_naive(a::MPOC64, b::MPOC64)

Contract two complex MPOs using naive (exact) algorithm.
"""
function contract_naive(a::MPOC64, b::MPOC64)
    lib = get_lib()
    ptr = ccall(
        Libdl.dlsym(lib, :t4a_mpo_c64_contract_naive),
        Ptr{Cvoid},
        (Ptr{Cvoid}, Ptr{Cvoid}),
        a.ptr,
        b.ptr,
    )
    return MPOC64(ptr)
end

"""
    contract_zipup(a::MPOC64, b::MPOC64; tolerance=1e-12, max_bond_dim=0)

Contract two complex MPOs using zip-up algorithm with compression.
"""
function contract_zipup(
    a::MPOC64,
    b::MPOC64;
    tolerance::Float64 = 1e-12,
    max_bond_dim::Int = 0,
)
    lib = get_lib()
    ptr = ccall(
        Libdl.dlsym(lib, :t4a_mpo_c64_contract_zipup),
        Ptr{Cvoid},
        (Ptr{Cvoid}, Ptr{Cvoid}, Cdouble, Csize_t),
        a.ptr,
        b.ptr,
        tolerance,
        max_bond_dim,
    )
    return MPOC64(ptr)
end

"""
    contract_fit(a::MPOC64, b::MPOC64; tolerance=1e-12, max_bond_dim=0, max_sweeps=0)

Contract two complex MPOs using variational fitting algorithm.
"""
function contract_fit(
    a::MPOC64,
    b::MPOC64;
    tolerance::Float64 = 1e-12,
    max_bond_dim::Int = 0,
    max_sweeps::Int = 0,
)
    lib = get_lib()
    ptr = ccall(
        Libdl.dlsym(lib, :t4a_mpo_c64_contract_fit),
        Ptr{Cvoid},
        (Ptr{Cvoid}, Ptr{Cvoid}, Cdouble, Csize_t, Csize_t),
        a.ptr,
        b.ptr,
        tolerance,
        max_bond_dim,
        max_sweeps,
    )
    return MPOC64(ptr)
end

"""
    contract(a::MPOC64, b::MPOC64; algorithm=:naive, tolerance=1e-12, max_bond_dim=0)

Contract two complex MPOs using the specified algorithm.

# Arguments
- `a, b`: MPOs to contract
- `algorithm`: Contraction algorithm (:naive, :zipup, or :fit)
- `tolerance`: Relative tolerance for truncation
- `max_bond_dim`: Maximum bond dimension (0 for unlimited)
"""
function contract(
    a::MPOC64,
    b::MPOC64;
    algorithm::Symbol = :naive,
    tolerance::Float64 = 1e-12,
    max_bond_dim::Int = 0,
)
    lib = get_lib()
    alg_cint = algorithm_to_cint(algorithm)
    ptr = ccall(
        Libdl.dlsym(lib, :t4a_mpo_c64_contract),
        Ptr{Cvoid},
        (Ptr{Cvoid}, Ptr{Cvoid}, Cint, Cdouble, Csize_t),
        a.ptr,
        b.ptr,
        alg_cint,
        tolerance,
        max_bond_dim,
    )
    return MPOC64(ptr)
end

# ============================================================================
# Copy
# ============================================================================

function Base.copy(mpo::MPOF64)
    lib = get_lib()
    ptr = ccall(
        Libdl.dlsym(lib, :t4a_mpo_f64_clone),
        Ptr{Cvoid},
        (Ptr{Cvoid},),
        mpo.ptr,
    )
    return MPOF64(ptr)
end

function Base.copy(mpo::MPOC64)
    lib = get_lib()
    ptr = ccall(
        Libdl.dlsym(lib, :t4a_mpo_c64_clone),
        Ptr{Cvoid},
        (Ptr{Cvoid},),
        mpo.ptr,
    )
    return MPOC64(ptr)
end

# ============================================================================
# Exports
# ============================================================================

export MPOF64, MPOC64
export zeros, constant, identity
export rank
export contract_naive, contract_zipup, contract_fit, contract

end # module MPO

File: python/tensor4all/src/tensor4all/mpo/_mpo.py
"""MPO (Matrix Product Operator) implementations."""

from __future__ import annotations

from enum import IntEnum
from typing import Literal, Sequence

from .._capi import get_lib, check_status


class ContractionAlgorithm(IntEnum):
    """Contraction algorithm for MPO operations."""

    NAIVE = 0
    ZIPUP = 1
    FIT = 2


def _algorithm_to_int(algorithm: str | ContractionAlgorithm) -> int:
    """Convert algorithm specification to integer."""
    if isinstance(algorithm, ContractionAlgorithm):
        return int(algorithm)
    if isinstance(algorithm, str):
        alg_map = {
            "naive": ContractionAlgorithm.NAIVE,
            "zipup": ContractionAlgorithm.ZIPUP,
            "fit": ContractionAlgorithm.FIT,
        }
        alg_lower = algorithm.lower()
        if alg_lower not in alg_map:
            raise ValueError(
                f"Unknown algorithm '{algorithm}'. "
                "Use 'naive', 'zipup', or 'fit'."
            )
        return int(alg_map[alg_lower])
    raise TypeError(f"Expected str or ContractionAlgorithm, got {type(algorithm)}")


class MPOF64:
    """Matrix Product Operator with float64 elements.

    An MPO represents a tensor with a Matrix Product structure,
    commonly used in quantum physics and tensor network computations.

    Examples
    --------
    >>> mpo = MPOF64.identity([2, 2])
    >>> len(mpo)
    2
    """

    __slots__ = ("_ptr", "_lib", "_ffi")

    def __init__(self, ptr, lib, ffi):
        """Initialize from a C pointer (internal use only)."""
        if ptr == ffi.NULL:
            raise ValueError("Failed to create MPOF64 (null pointer)")
        self._ptr = ptr
        self._lib = lib
        self._ffi = ffi

    def __del__(self):
        """Release the underlying C object."""
        if hasattr(self, "_ptr") and self._ptr != self._ffi.NULL:
            self._lib.t4a_mpo_f64_release(self._ptr)

    @classmethod
    def zeros(cls, site_dims: Sequence[tuple[int, int]]) -> MPOF64:
        """Create an MPO representing the zero operator.

        Parameters
        ----------
        site_dims : Sequence[tuple[int, int]]
            Site dimensions as (input_dim, output_dim) pairs.

        Returns
        -------
        MPOF64
            Zero MPO.
        """
        lib, ffi = get_lib()
        dims1 = ffi.new("size_t[]", [d[0] for d in site_dims])
        dims2 = ffi.new("size_t[]", [d[1] for d in site_dims])
        ptr = lib.t4a_mpo_f64_new_zeros(dims1, dims2, len(site_dims))
        return cls(ptr, lib, ffi)

    @classmethod
    def constant(
        cls, site_dims: Sequence[tuple[int, int]], value: float
    ) -> MPOF64:
        """Create an MPO representing a constant operator.

        Parameters
        ----------
        site_dims : Sequence[tuple[int, int]]
            Site dimensions as (input_dim, output_dim) pairs.
        value : float
            Constant value.

        Returns
        -------
        MPOF64
            Constant MPO.
        """
        lib, ffi = get_lib()
        dims1 = ffi.new("size_t[]", [d[0] for d in site_dims])
        dims2 = ffi.new("size_t[]", [d[1] for d in site_dims])
        ptr = lib.t4a_mpo_f64_new_constant(dims1, dims2, len(site_dims), value)
        return cls(ptr, lib, ffi)

    @classmethod
    def identity(cls, site_dims: Sequence[int]) -> MPOF64:
        """Create an identity MPO.

        Parameters
        ----------
        site_dims : Sequence[int]
            Site dimensions (same for input and output).

        Returns
        -------
        MPOF64
            Identity MPO.
        """
        lib, ffi = get_lib()
        dims = ffi.new("size_t[]", list(site_dims))
        ptr = lib.t4a_mpo_f64_new_identity(dims, len(site_dims))
        return cls(ptr, lib, ffi)

    def __len__(self) -> int:
        """Return the number of sites."""
        out_len = self._ffi.new("size_t*")
        status = self._lib.t4a_mpo_f64_len(self._ptr, out_len)
        check_status(status)
        return out_len[0]

    def rank(self) -> int:
        """Return the maximum bond dimension."""
        out_rank = self._ffi.new("size_t*")
        status = self._lib.t4a_mpo_f64_rank(self._ptr, out_rank)
        check_status(status)
        return out_rank[0]

    def copy(self) -> MPOF64:
        """Create a copy of this MPO."""
        ptr = self._lib.t4a_mpo_f64_clone(self._ptr)
        return MPOF64(ptr, self._lib, self._ffi)

    def contract_naive(self, other: MPOF64) -> MPOF64:
        """Contract with another MPO using naive algorithm.

        Parameters
        ----------
        other : MPOF64
            MPO to contract with.

        Returns
        -------
        MPOF64
            Contracted MPO.
        """
        ptr = self._lib.t4a_mpo_f64_contract_naive(self._ptr, other._ptr)
        return MPOF64(ptr, self._lib, self._ffi)

    def contract_zipup(
        self,
        other: MPOF64,
        *,
        tolerance: float = 1e-12,
        max_bond_dim: int = 0,
    ) -> MPOF64:
        """Contract with another MPO using zip-up algorithm.

        Parameters
        ----------
        other : MPOF64
            MPO to contract with.
        tolerance : float, optional
            Truncation tolerance.
        max_bond_dim : int, optional
            Maximum bond dimension (0 for unlimited).

        Returns
        -------
        MPOF64
            Contracted MPO.
        """
        ptr = self._lib.t4a_mpo_f64_contract_zipup(
            self._ptr, other._ptr, tolerance, max_bond_dim
        )
        return MPOF64(ptr, self._lib, self._ffi)

    def contract_fit(
        self,
        other: MPOF64,
        *,
        tolerance: float = 1e-12,
        max_bond_dim: int = 0,
        max_sweeps: int = 0,
    ) -> MPOF64:
        """Contract with another MPO using variational fitting.

        Parameters
        ----------
        other : MPOF64
            MPO to contract with.
        tolerance : float, optional
            Convergence tolerance.
        max_bond_dim : int, optional
            Maximum bond dimension (0 for unlimited).
        max_sweeps : int, optional
            Maximum number of sweeps (0 for default).

        Returns
        -------
        MPOF64
            Contracted MPO.
        """
        ptr = self._lib.t4a_mpo_f64_contract_fit(
            self._ptr, other._ptr, tolerance, max_bond_dim, max_sweeps
        )
        return MPOF64(ptr, self._lib, self._ffi)

    def contract(
        self,
        other: MPOF64,
        *,
        algorithm: str | ContractionAlgorithm = "naive",
        tolerance: float = 1e-12,
        max_bond_dim: int = 0,
    ) -> MPOF64:
        """Contract with another MPO using the specified algorithm.

        Parameters
        ----------
        other : MPOF64
            MPO to contract with.
        algorithm : str or ContractionAlgorithm, optional
            Algorithm to use: "naive", "zipup", or "fit".
        tolerance : float, optional
            Truncation tolerance.
        max_bond_dim : int, optional
            Maximum bond dimension (0 for unlimited).

        Returns
        -------
        MPOF64
            Contracted MPO.
        """
        alg_int = _algorithm_to_int(algorithm)
        ptr = self._lib.t4a_mpo_f64_contract(
            self._ptr, other._ptr, alg_int, tolerance, max_bond_dim
        )
        return MPOF64(ptr, self._lib, self._ffi)


class MPOC64:
    """Matrix Product Operator with complex128 elements.

    An MPO represents a tensor with a Matrix Product structure,
    commonly used in quantum physics and tensor network computations.

    Examples
    --------
    >>> mpo = MPOC64.identity([2, 2])
    >>> len(mpo)
    2
    """

    __slots__ = ("_ptr", "_lib", "_ffi")

    def __init__(self, ptr, lib, ffi):
        """Initialize from a C pointer (internal use only)."""
        if ptr == ffi.NULL:
            raise ValueError("Failed to create MPOC64 (null pointer)")
        self._ptr = ptr
        self._lib = lib
        self._ffi = ffi

    def __del__(self):
        """Release the underlying C object."""
        if hasattr(self, "_ptr") and self._ptr != self._ffi.NULL:
            self._lib.t4a_mpo_c64_release(self._ptr)

    @classmethod
    def zeros(cls, site_dims: Sequence[tuple[int, int]]) -> MPOC64:
        """Create an MPO representing the zero operator.

        Parameters
        ----------
        site_dims : Sequence[tuple[int, int]]
            Site dimensions as (input_dim, output_dim) pairs.

        Returns
        -------
        MPOC64
            Zero MPO.
        """
        lib, ffi = get_lib()
        dims1 = ffi.new("size_t[]", [d[0] for d in site_dims])
        dims2 = ffi.new("size_t[]", [d[1] for d in site_dims])
        ptr = lib.t4a_mpo_c64_new_zeros(dims1, dims2, len(site_dims))
        return cls(ptr, lib, ffi)

    @classmethod
    def constant(
        cls, site_dims: Sequence[tuple[int, int]], value: complex
    ) -> MPOC64:
        """Create an MPO representing a constant operator.

        Parameters
        ----------
        site_dims : Sequence[tuple[int, int]]
            Site dimensions as (input_dim, output_dim) pairs.
        value : complex
            Constant value.

        Returns
        -------
        MPOC64
            Constant MPO.
        """
        lib, ffi = get_lib()
        dims1 = ffi.new("size_t[]", [d[0] for d in site_dims])
        dims2 = ffi.new("size_t[]", [d[1] for d in site_dims])
        ptr = lib.t4a_mpo_c64_new_constant(
            dims1, dims2, len(site_dims), value.real, value.imag
        )
        return cls(ptr, lib, ffi)

    @classmethod
    def identity(cls, site_dims: Sequence[int]) -> MPOC64:
        """Create an identity MPO.

        Parameters
        ----------
        site_dims : Sequence[int]
            Site dimensions (same for input and output).

        Returns
        -------
        MPOC64
            Identity MPO.
        """
        lib, ffi = get_lib()
        dims = ffi.new("size_t[]", list(site_dims))
        ptr = lib.t4a_mpo_c64_new_identity(dims, len(site_dims))
        return cls(ptr, lib, ffi)

    def __len__(self) -> int:
        """Return the number of sites."""
        out_len = self._ffi.new("size_t*")
        status = self._lib.t4a_mpo_c64_len(self._ptr, out_len)
        check_status(status)
        return out_len[0]

    def copy(self) -> MPOC64:
        """Create a copy of this MPO."""
        ptr = self._lib.t4a_mpo_c64_clone(self._ptr)
        return MPOC64(ptr, self._lib, self._ffi)

    def contract_naive(self, other: MPOC64) -> MPOC64:
        """Contract with another MPO using naive algorithm.

        Parameters
        ----------
        other : MPOC64
            MPO to contract with.

        Returns
        -------
        MPOC64
            Contracted MPO.
        """
        ptr = self._lib.t4a_mpo_c64_contract_naive(self._ptr, other._ptr)
        return MPOC64(ptr, self._lib, self._ffi)

    def contract_zipup(
        self,
        other: MPOC64,
        *,
        tolerance: float = 1e-12,
        max_bond_dim: int = 0,
    ) -> MPOC64:
        """Contract with another MPO using zip-up algorithm.

        Parameters
        ----------
        other : MPOC64
            MPO to contract with.
        tolerance : float, optional
            Truncation tolerance.
        max_bond_dim : int, optional
            Maximum bond dimension (0 for unlimited).

        Returns
        -------
        MPOC64
            Contracted MPO.
        """
        ptr = self._lib.t4a_mpo_c64_contract_zipup(
            self._ptr, other._ptr, tolerance, max_bond_dim
        )
        return MPOC64(ptr, self._lib, self._ffi)

    def contract_fit(
        self,
        other: MPOC64,
        *,
        tolerance: float = 1e-12,
        max_bond_dim: int = 0,
        max_sweeps: int = 0,
    ) -> MPOC64:
        """Contract with another MPO using variational fitting.

        Parameters
        ----------
        other : MPOC64
            MPO to contract with.
        tolerance : float, optional
            Convergence tolerance.
        max_bond_dim : int, optional
            Maximum bond dimension (0 for unlimited).
        max_sweeps : int, optional
            Maximum number of sweeps (0 for default).

        Returns
        -------
        MPOC64
            Contracted MPO.
        """
        ptr = self._lib.t4a_mpo_c64_contract_fit(
            self._ptr, other._ptr, tolerance, max_bond_dim, max_sweeps
        )
        return MPOC64(ptr, self._lib, self._ffi)

    def contract(
        self,
        other: MPOC64,
        *,
        algorithm: str | ContractionAlgorithm = "naive",
        tolerance: float = 1e-12,
        max_bond_dim: int = 0,
    ) -> MPOC64:
        """Contract with another MPO using the specified algorithm.

        Parameters
        ----------
        other : MPOC64
            MPO to contract with.
        algorithm : str or ContractionAlgorithm, optional
            Algorithm to use: "naive", "zipup", or "fit".
        tolerance : float, optional
            Truncation tolerance.
        max_bond_dim : int, optional
            Maximum bond dimension (0 for unlimited).

        Returns
        -------
        MPOC64
            Contracted MPO.
        """
        alg_int = _algorithm_to_int(algorithm)
        ptr = self._lib.t4a_mpo_c64_contract(
            self._ptr, other._ptr, alg_int, tolerance, max_bond_dim
        )
        return MPOC64(ptr, self._lib, self._ffi)
